Add WeakRecipient (v2)

hoodie opened the pull request:
With this change you can now call `.recipient()` on `WeakAddr<A>` just
like on `Addr<A>`. In this case you get a `WeakRecipient<M>` that can
be upgraded to a `Recipient<M>`.

<!-- Thanks for considering contributing actix! -->
<!-- Please fill out the following to make our reviews easy. -->

## PR Type
<!-- What kind of change does this PR make? -->
<!-- Bug Fix / Feature / Refactor / Code Style / Other -->
INSERT_PR_TYPE


## PR Checklist
Check your PR fulfills the following:

<!-- For draft PRs check the boxes as you complete them. -->

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt


## Overview
Hi there,
I'm trying to convert my project to make greater use of `Recipient<M>` instead of `Addr<A>` since I believe it allows to loosen some tight coupling here and there. Unfortunately there is no equivalent to `WeakAddr<A>` for Recipients yet, so I cannot really use Recipients at the moment without running the risk of leaking objects.
Therefore I would like to suggest adding `WeakRecipients` to actix.

This PR supersedes #418, which was flawed. In that attempt I made the `Box` inside the `Recipient` an `Arc` and downgraded it.
you would call
```rust
let weak_rcp = my_actor.addr().recipient().downgrade();
```

That however meant  that you the `WeakRecipient` was invalidated immediately.
In this second attempt I'm creating the `WeakRecipient` from an existing `WeakAddr`. 
```rust
let weak_rcp = my_actor.addr().downgrade().recipient();
```

Thank you very much for your review.

hoodie commented:
@JohnTitor @robjtede ?

lucab commented on src/address/mod.rs:
```diff
@@ -329,6 +339,58 @@ where
+    }
+}
+
+/// The `Recipient` type allows to send one specific message to an
```
Spurious docstring (possibly a copy-paste leftover)?

hoodie replied:
whoops, yes indeed, I was shuffling files around a bit there, after I had moved recipient and weak_recipient into their own submodules and then reintegrated them. these files contain a bunch of structs at once and tons of trait implementations, I would at some point ask if we perhaps want to split them up a little.

tanks for the find!

lucab commented on src/address/channel.rs:
```diff
@@ -36,6 +36,13 @@ where
 
     fn connected(&self) -> bool;
 }
+pub trait WeakSender<M>: Send
```
Would `pub(crate)` be enough here?

lucab commented on src/address/mod.rs:
```diff
@@ -329,6 +339,58 @@ where
     }
 }
 
+/// A weakly referenced counterpart to `Recipient<M>`
```
As a user, I'd find some more docstring helpful here to explain what's the typical usecase of this weak recipient.
In particular, I found your mention of "risk of leaking objects" enlightening; it could mirrored and expanded in the docstring here.

hoodie replied:
this doc-string is a reflection of the one for `WeakAddr`. You are right about the usecase example, IDK if it fits in here, but maybe an example or integration test would be good.

JohnTitor replied:
The docs enhancement isn't absolutely necessary, but it'd be great if we could add a minimal example for it :)

lucab replied:
@hoodie thanks for the pair of examples, much appreciated!

lucab reviewed (commented):
(just some casual PR reading, I'm not a maintainer here)

hoodie commented:
Ping @JohnTitor 😀

JohnTitor reviewed (commented):
Hey, sorry for the delay, and thanks for pinging!
The approach should be reasonable and the change looks good to me. But I'm going to cc @Jonathas-Conceicao for further reviewing, if you're available now :)
Also, could you add an entry to the changelog?

hoodie commented on actix-derive/src/message.rs:
```diff
@@ -69,10 +69,10 @@ fn get_attribute_type_multiple(
     } else {
-        return Err(syn::Error::new_spanned(
+        Err(syn::Error::new_spanned(
             attr,
             format!("The correct syntax is #[{}(type, type, ...)]", name),
-        ));
+        ))
```
sure, will do 👍

hoodie commented:
@robjtede why did you merge `master` into this branch?

robjtede commented:
@hoodie repo has a linear history policy, so if this is accepted it needs to be based on HEAD for CI to check it properly before squashing

hoodie commented:
ah ok, so merging master in here is a actually better than rebasing onto master for this use case, get it

robjtede commented:
@hoodie either is appropriate and just as good, many people don't like rebasing and it's just a button on github to merge master in

Jonathas-Conceicao commented on src/address/channel.rs:
```diff
@@ -555,6 +562,22 @@ impl<A: Actor> WeakAddressSender<A> {
+    M::Result: Send,
+    M: Message + Send + 'static,
+{
+    fn upgrade(&self) -> Option<Box<dyn Sender<M> + Sync>> {
```
I believe we should have a doc note about when this `Option` would be `None`, either here or at the trait's definition.

hoodie replied:
sure, but probably at the definition of the trait and not here right?

Jonathas-Conceicao replied:
Both will be generated to the doc, but I think the trait it might be more noticeable. So if the `None` has a generic cause it would be better to have it there, if it's specific to the type's implementation we can have it here.

hoodie replied:
I added it on the trait, but it's `pub(crate)`, so it won't show up in any docs anyway.
Plus: the semantics are basically the same as for `Weak<T>::upgrade()`.
Alright by you @Jonathas-Conceicao ?

hoodie replied:
I added it on the trait, but it's pub(crate), so it won't show up in any docs anyway.
Plus: the semantics are basically the same as for Weak<T>::upgrade().
Alright by you @Jonathas-Conceicao ?

Jonathas-Conceicao commented on tests/test_address.rs:
```diff
@@ -100,7 +101,51 @@ impl Actor for WeakRunner {
+        drop(addr1);
+
+        ctx.run_later(Duration::new(0, 1000), move |_, _| {
+            if weak1.upgrade().is_none() {
+            } else {
```
```suggestion
            if weak1.upgrade().is_some() {
```

Jonathas-Conceicao reviewed (changes requested):
The feature proposal seams good overall, I left a couple of comments for some minor improvements.

hoodie commented:
@JohnTitor I have added an example for `WeakAddr` and `WeakRecipient`.
In the process I noticed that the signature of `Recipient::do_send()` is not consistent with the signature of `Addr::do_send()`, I believe the former should not return a `Result<...>`. I just opened #440 to track that.

hoodie commented:
Ok, since this has gotten a bunch of changes since I opened it originally I thought I'd give a quick recap since I know it's a bit hard to follow these things:

* [x] added `WeakRecipient`
* [x] added trait doc for `WeakAddressSender`
* [x] added two new examples
* [ ] make clippy happy

any other requests?

hoodie commented:
ping @Jonathas-Conceicao

Jonathas-Conceicao reviewed (approved):
lgtm, thanks for working on this!

JohnTitor commented:
Thanks all!
